File: cses/07_mathematics/1722_fibonacci-numbers/solution.py
import sys

MOD = 10 ** 9 + 7


def fib_pair(n):
    # Fast doubling: trả về cặp (F(n), F(n+1)) mod MOD.
    if n == 0:
        return (0, 1)
    # a = F(k), b = F(k+1) với k = n // 2
    a, b = fib_pair(n >> 1)
    # Hai đẳng thức fast doubling
    c = (a * ((2 * b - a) % MOD)) % MOD          # F(2k)
    d = (a * a + b * b) % MOD                     # F(2k+1)
    # Nếu bit hiện tại là 1 thì dịch thêm một bước Fibonacci
    if n & 1:
        return (d, (c + d) % MOD)                 # (F(2k+1), F(2k+2))
    else:
        return (c, d)                             # (F(2k), F(2k+1))


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])
    sys.stdout.write(str(fib_pair(n)[0] % MOD) + "\n")


if __name__ == "__main__":
    main()

File: cses/07_mathematics/1722_fibonacci-numbers/solution.cpp
#include <iostream>
#include <utility>

using namespace std;

namespace {
constexpr long long MOD = 1'000'000'007LL;

// Fast doubling: trả về cặp (F(n), F(n+1)) mod MOD.
// Đệ quy chia đôi n, dùng hai đẳng thức
//   F(2k)   = F(k) * (2*F(k+1) - F(k))
//   F(2k+1) = F(k)^2 + F(k+1)^2
pair<long long, long long> fibonacci_pair(unsigned long long n) {
    if (n == 0) {
        return {0, 1};
    }

    // a = F(k), b = F(k+1) với k = n / 2
    const auto [a, b] = fibonacci_pair(n >> 1ULL);
    // Cộng thêm MOD trước khi lấy mod vì (2*F(k+1) - F(k)) có thể âm
    const long long even = a * ((2 * b % MOD - a + MOD) % MOD) % MOD;  // F(2k)
    const long long odd = (a * a % MOD + b * b % MOD) % MOD;           // F(2k+1)

    // Nếu bit hiện tại là 1 thì dịch thêm một bước Fibonacci
    if (n & 1ULL) {
        return {odd, (even + odd) % MOD};
    }
    return {even, odd};
}
}  // namespace

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    unsigned long long n;
    cin >> n;
    cout << fibonacci_pair(n).first << '\n';
    return 0;
}
